File: harness/oracle/cases/sync-report.ts
/**
 * Tier-1 oracle case — the `quilltap sync` REPORT RENDERER (v4
 * `packages/quilltap/lib/sync-report.js`, added at `23da0b322`).
 *
 * Drives v4's REAL module and emits one NDJSON row per probe. The renderer is
 * deliberately pure — actions and a colour switch in, lines out — so the whole
 * surface is a table: the fixed 8/6 column widths, the capped path column, the
 * em dash a side-less action takes, the trailing slash on a folder, the
 * parenthetical's `sha …, size` tail, the summary's conditional clauses, and
 * the 0/1/2 exit code.
 *
 * Colour is recorded BOTH ways. v4 emits raw ANSI escapes and pads INSIDE them,
 * so the coloured form is not simply the plain form with escapes wrapped around
 * it — a port that reasoned about the widths rather than reproducing them would
 * be wrong only in the coloured half, which is the half a human ever sees.
 *
 * The module is plain CommonJS in the launcher package rather than something
 * `@/lib` resolves, so it is loaded by absolute path through `createRequire`
 * against the checkout's cwd (the memory note: a case outside the v4 tree
 * cannot bare-import a v4 dependency).
 *
 * Run from the v4 server checkout:
 *   cd ~/source/quilltap-server
 *   V5W=${V5W:-$HOME/source/quilltap-v5}
 *   npx tsx $V5W/harness/oracle/cases/sync-report.ts > /tmp/oracle-sync-report.ndjson
 */

import { createRequire } from 'module';
import { join } from 'path';

const require_ = createRequire(join(process.cwd(), 'noop.js'));
const report = require_(join(process.cwd(), 'packages/quilltap/lib/sync-report.js')) as {
  formatActionLine: (action: unknown, colour: boolean, pathWidth?: number) => string;
  formatActionLines: (actions: unknown[], colour: boolean) => string[];
  formatSummary: (summary: unknown, elapsedMs: number, dryRun: boolean) => string;
  exitCodeFor: (summary: unknown) => number;
  detailFor: (action: unknown) => string;
  displayPath: (action: unknown) => string;
  formatBytes: (n: number | undefined | null) => string;
};

const emit = (row: unknown) => process.stdout.write(JSON.stringify(row) + '\n');

// ---------------------------------------------------------------------------
// formatBytes
// ---------------------------------------------------------------------------

const BYTE_INPUTS: Array<number | null> = [
  0,
  1,
  512,
  1023,
  1024,
  1536,
  10240,
  1048575,
  1048576,
  1572864,
  1073741823,
  1073741824,
  1610612736,
  1099511627776,
  // The `.toFixed` rounding boundary: 1023.95 KB renders as "1024.0 KB", not
  // "1.0 MB" — the scale is chosen before the rounding.
  1048524,
  null,
];

for (const n of BYTE_INPUTS) {
  emit({ kind: 'format-bytes', id: String(n), n, out: report.formatBytes(n as number) });
}

// ---------------------------------------------------------------------------
// action lines
// ---------------------------------------------------------------------------

const SHA = 'abcdef0123456789'.repeat(4);

interface Action {
  kind: string;
  side: string | null;
  relativePath: string;
  entryKind: string;
  reason?: string;
  sha256?: string;
  sizeBytes?: number;
  outcome?: string;
  error?: string;
  description?: string;
  lastModified?: string;
}

const ACTIONS: Array<{ id: string; action: Action }> = [
  {
    id: 'create-disk-with-sha-and-size',
    action: {
      kind: 'create', side: 'disk', relativePath: 'chapters/01.md', entryKind: 'file',
      sha256: SHA, sizeBytes: 2048, outcome: 'planned',
    },
  },
  {
    id: 'modify-store-with-a-reason',
    action: {
      kind: 'modify', side: 'store', relativePath: 'chapters/03.md', entryKind: 'file',
      reason: 'disk newer by 2h 14m', sha256: SHA, sizeBytes: 900, outcome: 'planned',
    },
  },
  {
    id: 'mkdir-disk-a-folder-gets-a-slash',
    action: { kind: 'mkdir', side: 'disk', relativePath: 'lore/maps', entryKind: 'folder', outcome: 'planned' },
  },
  {
    id: 'rmdir-store-a-folder-with-a-reason',
    action: {
      kind: 'rmdir', side: 'store', relativePath: 'drafts', entryKind: 'folder',
      reason: 'deleted on disk since last sync', outcome: 'planned',
    },
  },
  {
    id: 'conflict-has-no-side',
    action: {
      kind: 'conflict', side: null, relativePath: 'ch.md', entryKind: 'file',
      reason: 'both sides changed; --prefer to resolve', outcome: 'skipped',
    },
  },
  {
    id: 'skip-has-no-side-either',
    action: {
      kind: 'skip', side: null, relativePath: 'a.md', entryKind: 'file',
      reason: '--direction to-disk', outcome: 'skipped',
    },
  },
  {
    id: 'touch-disk',
    action: {
      kind: 'touch', side: 'disk', relativePath: 'ch.md', entryKind: 'file',
      reason: 'mtime 2026-09-20T10:00:00.000Z', outcome: 'planned',
    },
  },
  {
    id: 'describe-disk',
    action: {
      kind: 'describe', side: 'disk', relativePath: 'lore/harbour.png', entryKind: 'file',
      description: 'A map.', outcome: 'planned',
    },
  },
  {
    id: 'delete-store',
    action: {
      kind: 'delete', side: 'store', relativePath: 'gone.md', entryKind: 'file',
      reason: 'deleted on disk since last sync', outcome: 'planned',
    },
  },
  {
    id: 'a-failure-takes-the-error-not-the-reason',
    action: {
      kind: 'create', side: 'disk', relativePath: 'bad.md', entryKind: 'file',
      reason: 'this reason is not printed', sha256: SHA, sizeBytes: 12,
      outcome: 'failed', error: 'EACCES: permission denied',
    },
  },
  {
    id: 'a-failure-with-no-error-falls-back-to-the-reason',
    action: {
      kind: 'create', side: 'disk', relativePath: 'bad2.md', entryKind: 'file',
      reason: 'a reason', outcome: 'failed',
    },
  },
  {
    id: 'a-create-with-no-sha-carries-no-tail',
    action: { kind: 'create', side: 'disk', relativePath: 'nosha.md', entryKind: 'file', outcome: 'planned' },
  },
  {
    id: 'a-create-with-a-sha-and-no-size',
    action: { kind: 'create', side: 'disk', relativePath: 'nosize.md', entryKind: 'file', sha256: SHA, outcome: 'planned' },
  },
  {
    id: 'a-touch-with-a-sha-carries-no-tail',
    action: { kind: 'touch', side: 'disk', relativePath: 'x.md', entryKind: 'file', sha256: SHA, sizeBytes: 5, outcome: 'planned' },
  },
  {
    id: 'an-unknown-kind-takes-no-colour',
    action: { kind: 'wibble', side: 'disk', relativePath: 'x.md', entryKind: 'file', reason: 'r', outcome: 'planned' },
  },
  {
    id: 'a-path-longer-than-the-action-column',
    action: {
      kind: 'modify', side: 'store', entryKind: 'file',
      relativePath: 'a/very/deeply/nested/directory/structure/with/a/long/name.md',
      reason: 'store newer by 3d 4h', sha256: SHA, sizeBytes: 1048576, outcome: 'planned',
    },
  },
  {
    id: 'a-path-with-no-detail-is-not-padded',
    action: { kind: 'touch', side: 'store', relativePath: 'short.md', entryKind: 'file', outcome: 'planned' },
  },
];

for (const { id, action } of ACTIONS) {
  for (const colour of [false, true]) {
    for (const pathWidth of [0, 20, 44]) {
      emit({
        kind: 'action-line',
        id: `${id}/colour=${colour}/width=${pathWidth}`,
        action,
        colour,
        pathWidth,
        out: report.formatActionLine(action, colour, pathWidth),
      });
    }
  }
  emit({
    kind: 'action-parts',
    id,
    action,
    detailFor: report.detailFor(action),
    displayPath: report.displayPath(action),
  });
}

// ---------------------------------------------------------------------------
// the whole plan, so the shared path column is measured
// ---------------------------------------------------------------------------

const PLANS: Array<{ id: string; actions: Action[] }> = [
  { id: 'empty', actions: [] },
  {
    id: 'mixed',
    actions: [
      ACTIONS[0].action, ACTIONS[1].action, ACTIONS[2].action, ACTIONS[4].action,
      ACTIONS[6].action, ACTIONS[9].action,
    ],
  },
  {
    // Every action carries a detail and one path is far past the cap, so the
    // column is the CAP and not the longest path.
    id: 'one-outlier-caps-the-column',
    actions: [ACTIONS[1].action, ACTIONS[15].action],
  },
  {
    // No action carries a detail, so the width stays 0 and nothing is padded.
    id: 'no-details-means-no-padding',
    actions: [ACTIONS[11].action, ACTIONS[16].action],
  },
];

for (const { id, actions } of PLANS) {
  for (const colour of [false, true]) {
    emit({
      kind: 'action-lines',
      id: `${id}/colour=${colour}`,
      actions,
      colour,
      out: report.formatActionLines(actions, colour),
    });
  }
}

// ---------------------------------------------------------------------------
// summary + exit code
// ---------------------------------------------------------------------------

interface Summary {
  created: number;
  modified: number;
  deleted: number;
  touched: number;
  described: number;
  conflicts: number;
  skipped: number;
  failed: number;
}

const zero: Summary = {
  created: 0, modified: 0, deleted: 0, touched: 0,
  described: 0, conflicts: 0, skipped: 0, failed: 0,
};

const SUMMARIES: Array<{ id: string; summary: Summary; elapsedMs: number }> = [
  { id: 'nothing', summary: { ...zero }, elapsedMs: 812 },
  { id: 'everything', summary: { created: 3, modified: 1, deleted: 2, touched: 4, described: 5, conflicts: 6, skipped: 7, failed: 8 }, elapsedMs: 1234 },
  { id: 'one-conflict-is-singular', summary: { ...zero, conflicts: 1 }, elapsedMs: 50 },
  { id: 'two-conflicts-are-plural', summary: { ...zero, conflicts: 2 }, elapsedMs: 50 },
  { id: 'only-created', summary: { ...zero, created: 3 }, elapsedMs: 100 },
  { id: 'only-failed', summary: { ...zero, failed: 1 }, elapsedMs: 100 },
  { id: 'failed-and-conflicted', summary: { ...zero, failed: 1, conflicts: 1 }, elapsedMs: 100 },
  // The seconds render through `toFixed(1)`, so the rounding is part of it.
  { id: 'sub-millisecond', summary: { ...zero, touched: 1 }, elapsedMs: 0 },
  { id: 'rounds-up', summary: { ...zero, touched: 1 }, elapsedMs: 1950 },
  { id: 'rounds-to-a-tenth', summary: { ...zero, touched: 1 }, elapsedMs: 1249 },
  { id: 'long-run', summary: { ...zero, created: 1 }, elapsedMs: 3_723_000 },
];

for (const { id, summary, elapsedMs } of SUMMARIES) {
  for (const dryRun of [false, true]) {
    emit({
      kind: 'summary',
      id: `${id}/dryRun=${dryRun}`,
      summary,
      elapsedMs,
      dryRun,
      out: report.formatSummary(summary, elapsedMs, dryRun),
    });
  }
  emit({ kind: 'exit-code', id, summary, out: report.exitCodeFor(summary) });
}

File: crates/quilltap-cli/src/sync_report.rs
//! Rendering for `quilltap sync` — port of v4
//! `packages/quilltap/lib/sync-report.js` (`23da0b322`).
//!
//! Deliberately pure — actions and a colour switch in, lines out — so the
//! report's shape can be tested without a server, a store, or a directory, the
//! way `docker-mounts` is.
//!
//! ## The shape of a line
//!
//! ```text
//!     modify   store  chapters/03.md                (disk newer by 2h 14m)
//!     ^        ^      ^                             ^
//!     action   side   path                          why
//! ```
//!
//! The first two columns are fixed width so the paths line up under each other,
//! because the thing an operator actually reads down is the path column. The
//! side is the side that CHANGES: `modify store` means the store is rewritten
//! from disk, which is the opposite of the intuition some people bring and so is
//! worth being unambiguous about.
//!
//! Advisory text — warnings, the summary — goes to stderr in the caller, so
//! stdout stays greppable.
//!
//! **The shapes here are LOOSE on purpose.** v4's renderer is plain JS reading
//! whatever JSON the server sent: an action kind it does not recognize simply
//! takes no colour, and a missing key reads as `undefined`. A typed enum would
//! refuse that payload instead of rendering it, so `kind` and `side` stay
//! strings and every optional field is an `Option`. They are built out of
//! `serde_json::Value` by hand for the same reason the rest of this crate does
//! (`db_characters.rs`): the crate links `serde_json` and not `serde`, and a
//! JSON-shaped reader is what v4 is.

use serde_json::Value;

use crate::nodefmt::format_bytes as node_format_bytes;
use quilltap_core::jsstr::utf16_len;

const RESET: &str = "\x1b[0m";
const DIM: &str = "\x1b[2m";
const GREEN: &str = "\x1b[32m";
const RED: &str = "\x1b[31m";
const YELLOW: &str = "\x1b[33m";

/// Column widths. `conflict` is the longest action word at 8.
const ACTION_WIDTH: usize = 8;
const SIDE_WIDTH: usize = 6;

/// Longest path that carries a detail, capped so one outlier cannot ruin the
/// column.
const MAX_PATH_COLUMN: usize = 44;

/// One action as the server sent it.
#[derive(Debug, Clone, Default)]
pub struct ReportAction {
    pub kind: String,
    pub side: Option<String>,
    pub relative_path: String,
    pub entry_kind: Option<String>,
    pub reason: Option<String>,
    pub sha256: Option<String>,
    pub size_bytes: Option<f64>,
    pub outcome: Option<String>,
    pub error: Option<String>,
}

/// `obj[key]` as a JS string, or `None` for absent / `null` / a non-string —
/// which is what every read below then tests for truthiness anyway.
fn str_at(value: &Value, key: &str) -> Option<String> {
    value.get(key).and_then(Value::as_str).map(str::to_string)
}

fn num_at(value: &Value, key: &str) -> Option<f64> {
    value.get(key).and_then(Value::as_f64)
}

fn int_at(value: &Value, key: &str) -> i64 {
    value.get(key).and_then(Value::as_i64).unwrap_or(0)
}

impl ReportAction {
    pub fn from_value(value: &Value) -> Self {
        Self {
            kind: str_at(value, "kind").unwrap_or_default(),
            side: str_at(value, "side"),
            relative_path: str_at(value, "relativePath").unwrap_or_default(),
            entry_kind: str_at(value, "entryKind"),
            reason: str_at(value, "reason"),
            sha256: str_at(value, "sha256"),
            size_bytes: num_at(value, "sizeBytes"),
            outcome: str_at(value, "outcome"),
            error: str_at(value, "error"),
        }
    }
}

#[derive(Debug, Clone, Copy, Default)]
pub struct ReportSummary {
    pub created: i64,
    pub modified: i64,
    pub deleted: i64,
    pub touched: i64,
    pub described: i64,
    pub conflicts: i64,
    pub skipped: i64,
    pub failed: i64,
}

impl ReportSummary {
    pub fn from_value(value: &Value) -> Self {
        Self {
            created: int_at(value, "created"),
            modified: int_at(value, "modified"),
            deleted: int_at(value, "deleted"),
            touched: int_at(value, "touched"),
            described: int_at(value, "described"),
            conflicts: int_at(value, "conflicts"),
            skipped: int_at(value, "skipped"),
            failed: int_at(value, "failed"),
        }
    }
}

/// The whole report, as the CLI reads it back off the wire.
#[derive(Debug, Clone, Default)]
pub struct SyncReportView {
    pub actions: Vec<ReportAction>,
    pub summary: ReportSummary,
    pub warnings: Vec<String>,
    pub elapsed_ms: f64,
    pub dry_run: bool,
}

impl SyncReportView {
    pub fn from_value(value: &Value) -> Self {
        Self {
            actions: value
                .get("actions")
                .and_then(Value::as_array)
                .map(|a| a.iter().map(ReportAction::from_value).collect())
                .unwrap_or_default(),
            summary: ReportSummary::from_value(value.get("summary").unwrap_or(&Value::Null)),
            // v4 `report.warnings || []`.
            warnings: value
                .get("warnings")
                .and_then(Value::as_array)
                .map(|a| {
                    a.iter()
                        .map(|w| w.as_str().unwrap_or_default().to_string())
                        .collect()
                })
                .unwrap_or_default(),
            elapsed_ms: num_at(value, "elapsedMs").unwrap_or(0.0),
            dry_run: value
                .get("dryRun")
                .and_then(Value::as_bool)
                .unwrap_or(false),
        }
    }
}

/// v4's `ACTION_COLOURS` table. An unknown kind takes `''` — no colour, and no
/// reset either.
fn action_colour(kind: &str) -> &'static str {
    match kind {
        "create" | "mkdir" | "modify" | "describe" => GREEN,
        "delete" | "rmdir" => YELLOW,
        "touch" | "skip" => DIM,
        "conflict" => RED,
        _ => "",
    }
}

/// JS `text.length >= width ? text : text + ' '.repeat(width - text.length)` —
/// `String#length` is UTF-16 code units, and the ANSI escapes v4 pads INSIDE
/// count toward it.
fn pad(text: &str, width: usize) -> String {
    let len = utf16_len(text);
    if len >= width {
        text.to_string()
    } else {
        format!("{text}{}", " ".repeat(width - len))
    }
}

/// v4 `formatBytes`.
///
/// Measured at the pin: the body is byte-identical to `docs-commands.js`'s,
/// which [`crate::nodefmt::format_bytes`] already ports (the same four branches,
/// the same `toFixed` digits, the same division spellings). The ONLY difference
/// is the absent-value guard this one adds, so the port reuses the existing
/// kernel rather than minting a fourth copy.
pub fn format_bytes(n: Option<f64>) -> String {
    match n {
        None => String::new(),
        Some(n) => node_format_bytes(n),
    }
}

/// v4 `detailFor` — the parenthetical after a path: whatever the engine said,
/// plus the sha and size on a line that put bytes somewhere.
pub fn detail_for(action: &ReportAction) -> String {
    let mut parts: Vec<String> = Vec::new();
    // v4 `action.outcome === 'failed' && action.error` — JS truthiness, so an
    // empty error string falls through to the reason.
    let failed_with_error = action.outcome.as_deref() == Some("failed")
        && action.error.as_deref().is_some_and(|e| !e.is_empty());
    if failed_with_error {
        parts.push(format!("FAILED: {}", action.error.as_deref().unwrap_or("")));
    } else if let Some(reason) = action.reason.as_deref().filter(|r| !r.is_empty()) {
        parts.push(reason.to_string());
    }
    if (action.kind == "create" || action.kind == "modify")
        && action.sha256.as_deref().is_some_and(|s| !s.is_empty())
    {
        let sha = action.sha256.as_deref().unwrap_or("");
        let size = format_bytes(action.size_bytes);
        let tail = if size.is_empty() {
            String::new()
        } else {
            format!(", {size}")
        };
        // `sha256.slice(0, 4)` on a JS string — UTF-16 units; a sha is hex, so
        // this is four bytes in practice.
        let short: String = sha.chars().take(4).collect();
        parts.push(format!("sha {short}\u{2026}{tail}"));
    }
    parts.join("; ")
}

/// v4 `displayPath` — a trailing slash marks a folder, so `drafts/` and a file
/// called `drafts` are not the same line.
pub fn display_path(action: &ReportAction) -> String {
    if action.entry_kind.as_deref() == Some("folder") {
        format!("{}/", action.relative_path)
    } else {
        action.relative_path.clone()
    }
}

/// v4 `formatActionLine`.
pub fn format_action_line(action: &ReportAction, colour: bool, path_width: usize) -> String {
    let tint = if colour {
        if action.outcome.as_deref() == Some("failed") {
            RED
        } else {
            action_colour(&action.kind)
        }
    } else {
        ""
    };
    // v4 `const reset = tint ? RESET : ''` — an unknown kind under colour takes
    // neither.
    let reset = if tint.is_empty() { "" } else { RESET };
    // v4 `action.side || '—'` — JS truthiness, so `null` AND an empty string
    // both take the em dash.
    let side = match action.side.as_deref() {
        Some(s) if !s.is_empty() => s,
        _ => "\u{2014}",
    };
    let detail = detail_for(action);
    let shown_path = if detail.is_empty() {
        display_path(action)
    } else {
        pad(&display_path(action), path_width)
    };
    let head = format!(
        "{tint}{}{reset} {} {shown_path}",
        pad(&action.kind, ACTION_WIDTH),
        pad(side, SIDE_WIDTH),
    );
    if detail.is_empty() {
        return head;
    }
    if colour {
        format!("{head}  {DIM}({detail}){RESET}")
    } else {
        format!("{head}  ({detail})")
    }
}

/// v4 `formatActionLines` — one line per action, with the path column measured
/// once across the whole plan so a single very long path does not push every
/// other line's detail off the screen.
pub fn format_action_lines(actions: &[ReportAction], colour: bool) -> Vec<String> {
    let widest = actions.iter().fold(0usize, |widest, action| {
        if detail_for(action).is_empty() {
            widest
        } else {
            widest.max(utf16_len(&display_path(action)))
        }
    });
    let path_width = MAX_PATH_COLUMN.min(widest);
    actions
        .iter()
        .map(|action| format_action_line(action, colour, path_width))
        .collect()
}

/// v4 `formatSummary` — `3 created, 1 modified, … — 0.8 s`, or a plain "nothing
/// to do".
pub fn format_summary(summary: &ReportSummary, elapsed_ms: f64, dry_run: bool) -> String {
    let mut bits: Vec<String> = Vec::new();
    // v4 tests each count for JS truthiness, so a zero is simply absent. The
    // ORDER is v4's, and it is not the summary struct's: `skipped` comes before
    // `conflicts`, and `failed` is last.
    if summary.created != 0 {
        bits.push(format!("{} created", summary.created));
    }
    if summary.modified != 0 {
        bits.push(format!("{} modified", summary.modified));
    }
    if summary.deleted != 0 {
        bits.push(format!("{} deleted", summary.deleted));
    }
    if summary.touched != 0 {
        bits.push(format!("{} touched", summary.touched));
    }
    if summary.described != 0 {
        bits.push(format!("{} described", summary.described));
    }
    if summary.skipped != 0 {
        bits.push(format!("{} skipped", summary.skipped));
    }
    if summary.conflicts != 0 {
        bits.push(format!(
            "{} conflict{}",
            summary.conflicts,
            if summary.conflicts == 1 { "" } else { "s" }
        ));
    }
    if summary.failed != 0 {
        bits.push(format!("{} failed", summary.failed));
    }

    let seconds = format!(
        "{} s",
        quilltap_core::jsnum::to_fixed(elapsed_ms / 1000.0, 1)
    );
    if bits.is_empty() {
        return if dry_run {
            format!("Nothing to do \u{2014} {seconds}")
        } else {
            format!("Already in step \u{2014} {seconds}")
        };
    }
    let prefix = if dry_run { "Would do: " } else { "" };
    format!("{prefix}{} \u{2014} {seconds}", bits.join(", "))
}

/// v4 `exitCodeFor` — the process exit code a report earns.
///
///   0  clean
///   1  something failed outright
///   2  at least one conflict is still unresolved
///
/// `--dry-run` uses the same codes, so a script can gate on a clean plan before
/// it lets a real run proceed.
pub fn exit_code_for(summary: &ReportSummary) -> i32 {
    if summary.failed > 0 {
        return 1;
    }
    if summary.conflicts > 0 {
        return 2;
    }
    0
}
